Chunk passing

Leporacanthicus opened the pull request:
Fix the fact that chunk-size is not passed to the WSLoop. This is pretty much the smallest fix I could come up with, it duplicates a loop from the code further down, but moving the entire loop up, and then setting up various variables to fetch out all the other attributes is probably more complicated and not beneficial. As I understand it, the number of elements in the wsLoopClauseList is relatively small in all valid cases, so repeating the loop is not a big overhead.

SouraVX commented on flang/test/Lower/OpenMP/omp-wsloop-chunks.f90:
```diff
@@ -0,0 +1,97 @@
+! This test checks lowering of OpenMP DO Directive(Worksharing).
```
NIT: Can you make this comment a bit more specific ?

Leporacanthicus replied:
Updated.

SouraVX commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -369,6 +369,22 @@ static void genOMP(Fortran::lower::AbstractConverter &converter,
+  for (const auto &clause : wsLoopOpClauseList.v) {
+    if (const auto &scheduleClause =
+            std::get_if<Fortran::parser::OmpClause::Schedule>(&clause.u)) {
+      if (const auto &chunkExpr{
```
NIT: Please avoid using f18 style `{}`, For lowering code we use MLIR coding style :)

Leporacanthicus replied:
Fixed

SouraVX commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -369,6 +369,22 @@ static void genOMP(Fortran::lower::AbstractConverter &converter,
+      if (const auto &chunkExpr{
+              std::get<std::optional<Fortran::parser::ScalarIntExpr>>(
+                  scheduleClause->v.t)}) {
+        if (const auto v{Fortran::semantics::GetIntValue(*chunkExpr)}) {
```
NIT: Same here!

Leporacanthicus replied:
Fxed.

SouraVX reviewed (approved):
Thanks! @Leporacanthicus for the patch :)

SouraVX commented:
>This is pretty much the smallest fix I could come up with, it duplicates a loop from the code further down, but moving the entire loop up, and then setting up various variables to fetch out all the other attributes is probably more complicated and not beneficial. As I understand it, the number of elements in the wsLoopClauseList is relatively small in all valid cases, so repeating the loop is not a big overhead.


👍

kiranchandramohan reviewed (commented):
1) Consider renaming the subject of this PR to something like, [OpenMP] Pass the chunk size to MLIR while lowering from parse-tree.
2) chunk-size can be a positive expression. Add more tests with chunk size as a variable, variable expression, constant expression (like 2+2).
3) The test included can fail when fir-dev is rebased since the pretty-printer will arrive. I suggest simplifying the test to test the specific thing that we want to test. i.e for the included test, the chunk size 4 appears in the LLVM IR.

Leporacanthicus commented:
> 1. Consider renaming the subject of this PR to something like, [OpenMP] Pass the chunk size to MLIR while lowering from parse-tree.

I have renamed the commit itself. Not sure how you rename a PR?

> 2. chunk-size can be a positive expression. Add more tests with chunk size as a variable, variable expression, constant expression (like 2+2).

Added variable and constant expression.

> 3. The test included can fail when fir-dev is rebased since the pretty-printer will arrive. I suggest simplifying the test to test the specific thing that we want to test. i.e for the included test, the chunk size 4 appears in the LLVM IR.

Removed the MLIR and FIR testing.

clementval commented:
> I have renamed the commit itself. Not sure how you rename a PR?
There is a "edit" button on the top-right corner.

Leporacanthicus commented:
> > I have renamed the commit itself. Not sure how you rename a PR?
> > There is a "edit" button on the top-right corner.

As simple as that! :)
Thanks!

kiranchandramohan reviewed (approved):
LGTM.
Nit: Title of the commit: Spelling OpneMP -> OpenMP

Leporacanthicus commented:
> LGTM.
> Nit: Title of the commit: Spelling OpneMP -> OpenMP

Fixed...

schweitzpgi commented on flang/lib/Lower/OpenMP.cpp:
```diff
@@ -22,6 +22,7 @@
 #include "flang/Semantics/tools.h"
 #include "mlir/Dialect/OpenMP/OpenMPDialect.h"
 #include "llvm/Frontend/OpenMP/OMPConstants.h"
+#include <iostream>
```
IIRC, iostream is a no-no in LLVM.

schweitzpgi replied:
The good news is that this line looks unnecessary.

Leporacanthicus replied:
Oops, remnants of debugging. Update coming in a bit!

kiranchandramohan commented:
@schweitzpgi is this patch OK with you?

schweitzpgi reviewed (approved):
LGTM
